mbbsd/email_challenge.cc

```diff
@@ -0,0 +1,203 @@
+extern "C" {
+#include "bbs.h"
+#include "daemons.h"
+}
+
+#ifdef USE_VERIFYDB_ACCOUNT_RECOVERY
+
+# ifndef USE_VERIFYDB
+#   error "USE_VERIFYDB_ACCOUNT_RECOVERY requires USE_VERIFYDB"
+# endif
+
+#include <optional>
+#include <string>
+
+#include "verifydb.h"
+#include "verifydb.fbs.h"
+
+namespace emailchallenge {
+
+constexpr int kMaxErrCnt = 3;
+constexpr size_t kCodeLen = 30;
+
+static bool LoadVerifyDbEmail(const userec_t *user, std::vector<std::string> &all_emails);
+static std::optional<std::string> NormalizeEmail(const std::string &email);
+static std::string GenCode(size_t len);
+static bool SendChallengeCode(const std::string &email, const std::string &code, const std::string &prompt, const std::string &ip, const std::string &filename);
+static void UserErrorExit();
+
+// EmailChallenge
+//
+// params:
+//     email: user-input-email
+//     user: user
+//     y: starting y onscreen
+//     prompt: prompt for the email title. (prompt [code] @ IP ip)
+//     ip: ip for the email title.
+//     filename: filename for the email template.
+//
+// return:
+//     bool: true: success / false: faled
+//     out_y: cursor-y after returning from the function.
+bool EmailChallenge(const std::string &email, const userec_t *user, const int y, const std::string &prompt, const std::string &ip, const std::string &filename, int *out_y) {
+  std::vector<std::string> all_emails = {};
+  std::optional<std::string> normalized_email_opt = std::nullopt;
+  std::string normalized_email = {};
+
+  // load verify-db email (regemail)
+  if(!LoadVerifyDbEmail(user, all_emails)) {
+    vmsg("�t�ο�~�A�еy�ԦA�աC");
+    exit(0);
+  }
+
+  // load contact email
+#ifdef USEREC_EMAIL_IS_CONTACT
+  normalized_email_opt = NormalizeEmail(user->email);
+  if(normalized_email_opt) {
+    all_emails.push_back(normalized_email_opt.value());
+  }
+#endif
+
+  int current_y = y;
+  auto code = GenCode(kCodeLen);
+  mvprints(current_y++, 0, "�t�γB�z��...");
+  doupdate();
+
+  // Check and send challenge code. Don't exit if email doesn't match, so user
+  // can't guess email.
+  bool email_matches = false;
+  normalized_email_opt = NormalizeEmail(email);
+  if(normalized_email_opt) {
+    //in case that normalized_email is a copy of email.
+    normalized_email = normalized_email_opt.value();
+    for (const auto &each_email : all_emails) {
+      if (each_email == normalized_email) {
+        email_matches = true;
+      }
+    }
+  }
+
+  if (email_matches) {
+    // We only want to send email if the user input the matching address.
+    // Silently fail if email is not matching, so that user cannot guess other
+    // user's email address.
```

Comment:
I'm not sure what you want to achieve with this. The user's email is already visible through the user info menu, even the email stored in the verification database.

Reply:
I wrote email_challenge.cc based on recovery.cc

This is just trying to minimize the difference between recovery.cc and email_challenge.cc:

https://github.com/ptt/pttbbs/blob/master/mbbsd/recover.cc#L227

Reply:
`all_emails` includes the contact email and the "original verified email".

It is possible that some user's contact email got updated by some bad guy, 
but the user can still use the "original verified email" to do the account-related operations (retrieve the account / change the contact-email again.)

Reply:
`all_emails` includes "contact email" and "original verified email".

It's possible that some user's contact email got updated by some bad guy,
but the user can still use "original verified email" to do account-related operations (retrieve the account / change the email.)

Reply:
What I meant was the original verified email is visible via (U)ser -> (I)nfo -> (V), so the "prevent bad user from guessing" is moot.

Plus you already copied the implementation. What's the point of minimizing the difference? It's not like you are reusing it.

Reply:
Re: 1.  But this is a separate function, i.e. 2FA _before_ changing password / email, so the user (legit or otherwise) already has access to said info. Why bother having the user input it again? Just send the challenge email to both the verified email and contact email.

Re: 2. Either merge them (which I doubt would work or look pretty), or just write functions that suite your need. Duplicating it is the worst choice.

Reply:
@wens 

Thanks for the reply.

1. I didn't think of this alternative setup. I'll update to this setup.
2. I'll merge these two functions to simplify the code.